### DISCUSSION TITLE
Gracefully Handling 'No Data' Requests

### DISCUSSION THREAD

**Discussion by cjolson64** [ERDDAP/erddap#209] [General]
I've been building a basic web frontend for some datasets I have in ERDDAP and am running into a bit of messiness with some of my GET requests.  When I run a query where there are no results, say because there is no data between the date ranges provided or some other type of filtering, I am getting a 404 error.  I think 404 makes sense if the dataset cannot be found, but I feel like this should still give a 200 just with zero results, definitely open to having my mind changed though.

The real issue though is that whenever I get a 404 in this situation it comes back without CORS headers.  Has anyone else run into this situation?

**Comment by thogar-computer** [comment]
hiya @cjolson64 

we have been building out a frontend over our ERDDAP instance too. Current version is 
[https://linkedsystems.uk/dataportal/](https://linkedsystems.uk/dataportal/) severing content from [https://linkedsystems.uk/erddap/index.html](https://linkedsystems.uk/erddap/index.html) 

we have also started making smaller custom dashboards for smaller collections of data [sealevel](https://www.bodc.ac.uk/data/hosted_data_systems/sea_level/uk_tide_gauge_network/dashboard/)

I'll mention this to one of our developers leading on the dashboard. @SarahSidders

but for what its worth, my option is that the returning of a 404 for when data cannot be found is correct, both in terms of the restful standard and in how I would expect an API to work. 

the CORS issue is interesting, do you have control over the erddap instance or can you request changes from your web admin?

**Reply by ChrisJohnNOAA** [reply]
I do think 404 is the most appropriate response in the no data found case. You should be able to parse the included message to determine whether it's a bad query:

Error {
    code=404;
    message="Not Found: Your query produced no matching results. Query error: For variable=longitude axis#2=longitude Constraint=\"[(-6.23721928665664):1:(-6.23721928665664)]\": Start=\"-6.23721928665664\" is greater than the axis maximum=-56.23721928665664 (and even -56.212771548689716).";
}

or bad dataset:

Error {
    code=404;
    message="Not Found: Currently unknown datasetID=asdfasdf";
}

As for the headers issue, assuming you're talking about Access-Control-* headers, Those are handled outside of ERDDAP itself (so in Tomcat or other hosting server), or at least on a server by server basis by the admins. For the Tomcat instructions you can read more [in their documentation.](https://tomcat.apache.org/tomcat-10.0-doc/config/filter.html#CORS_Filter)

**Comment by srstsavage** [comment]
I do wish a different HTTP status code had originally been chosen for this situation (requested dataset found, but no data matching query parameters), maybe 400 BAD REQUEST. Changing it or making it configurable now would be very painful though.

For the CORS issue, if you're using the `axiom/docker-erddap` Docker image the Tomcat `org.apache.catalina.filters.CorsFilter` is enabled by default.

https://tomcat.apache.org/tomcat-9.0-doc/config/filter.html

The `CorsFilter` configuration exists in `web.conf`, which is imported from the `Unidata/tomcat-docker` image.

https://github.com/Unidata/tomcat-docker/blob/latest/web.xml#L629-L641

I'm not seeing an option to enable CORS headers on error responses for `org.apache.catalina.filters.CorsFilter`, so if you're using it, you may want to set env var `DISABLE_CORS=1` on your ERDDAP Docker container and stand up a CORS proxy in front of ERDDAP that your frontend can hit. `nginx` can do it quite well, as can a number of proxies built for this exact use case.

**Comment by srstsavage** [comment]
Follow up thought on this. I've been seeing more and more web server configuration to enable automated banning of perceived bad actors based on detecting patterns of many 404 requests (i.e. a scanner requests `/password.txt`, `/microsoft-product.conf`, `/api/run-as-root`, etc looking for vulnerabilities). Auto-ban policies like this will probably continue to increase as agentic attacks become more common. Usually this is a good heuristic to filter out malicious scanners, but a legitimate ERDDAP client could easily trigger this as well if it was polling datasets for updates since any dataset without data in the requested time range would return 404. It might be worth revisiting the possibility of adding ERDDAP config to change the return code for "no date in filter range" and distinguish it from "dataset/page not found at all."

**Reply by ChrisJohnNOAA** [reply]
I'm definitely open to an option to make ERDDAP work well with auto-ban configurations.

Just to make sure I understand, we're talking about something like [fail2ban](https://github.com/fail2ban/fail2ban), which relies on the status code to determine bad actors and puts them (temporarily) in "jail". So any series of requests that get 400 (or 500?) responses would potentially trigger an ban. My 2 cents here is that it may differ by server admin what should be a potentially ban worthy request.

I see a few types of requests:
1) Good requests that return data or a page
2) Data requests that happen to have no data given otherwise valid filters
3) Data requests that have invalid, but reasonable filters (data for today but the dataset isn't realtime)
4) Data requests with unreasonable filters/variables that don't exist
5) Malicious/probing/scanner requests

It may be tricky differentiating 2, 3, and 4 in code (particularly 3 and 4). I think 4 (when we can detect it) and 5 should always return errors. I could see differences of opinion on 2 and 3.

Related to this topic, I've been planning to add one or more [Servlet Filters](https://www.oracle.com/java/technologies/filters.html) to handle things like request spam (there's currently a version of this logic in Erddap::doGet). I could easily see that expanding to support ERDDAP level bad actor (temporary?) banning.

**Reply by srstsavage** [reply]
> we're talking about something like [fail2ban](https://github.com/fail2ban/fail2ban)

Yes, I'm thinking of exactly fail2ban!

Based on your categories above, my opinion is that ideal/"proper" behavior would be:

1. Good requests that return data or a page - return 200
1. Data requests that happen to have no data given otherwise valid filters - return 200
1. Data requests that have invalid, but reasonable filters (data for today but the dataset isn't realtime) - return 200
1. Data requests with unreasonable filters/variables that don't exist - return 400
1. Malicious/probing/scanner requests - return 404

1, 4, and 5 seem straightforward, 2 and 3 probably controversial and maybe wrong. In my mind the HTTP resource is the dataset, so if the requested dataset exists the response should never be 404. 204 No Content is tempting for 2 and 3 but no content is allowed in the response. Kinda surprising that there's no HTTP code that seems especially well matched to this situation :thinking:

**Comment by CaraWilson** [comment]
On a slightly different note, we (CoastWatch West Coast/PolarWatch) were wondering if there could be a way to implement notification to users if they are accessing a dataset that has been removed, deprecated, or renamed (dataset ID change). It would pop up like an error message but would be different than a 404 or other error code. Rather, something customizable would be displayed that conveys that the dataset no longer available at this location, with a link leading to a page more information. Some use cases are below:
1) Notify users that a new version of a dataset is available. 
2) When the source data for a dataset becomes unavailable, we can explain why and point a user to alternate datasets.

**Reply by ChrisJohnNOAA** [reply]
I've made a new issue for this: #409 because while it's related to errors on dataset requests, the implementation logic will be completely different.

**Comment by CaraWilson** [comment]
Thanks Chris!
